### ISSUE TITLE
Migration causes failure if the specific code already exists on the engine code is migrating to

### ISSUE THREAD

**Issue by marusshi** [karthikp-ubc/ThingsJS#22]
Code Engine keeps track of all codes it is running in a field called ```codes```, a mapping of the code name to Code object.
When code migrates from engine n<sub>0</sub> to n<sub>1</sub>, n<sub>1</sub>'s ```codes``` field is updated to the code recovered from the snapshot
```
		var code = Code.fromSnapshot(kwargs.snapshot);
		self.codes[code.name] = code;
```
This will cause problems in an example like this:
1. Code Engine n<sub>0</sub> is running Foo.js
2. Code Engine n<sub>1</sub> is running Foo.js
3. n<sub>1</sub> migrates its Foo.js instance to n<sub>0</sub>
4. n<sub>0</sub>'s original Code object for Foo.js is overwritten and its reference is lost

This will specifically cause problems for the Scheduler because it assumes n<sub>0</sub> still has reference to its original Foo.js. If the next schedule is computed where n<sub>0</sub> should migrate its original Foo.js, an error like this will be thrown:
```
TypeError: Cannot read property 'snapshot' of undefined at Object.migrate_code (ThingsJS\lib\core\CodeEngine.js:269:69)
```

**Comment by marusshi** [comment]
I found a workaround for the SchedulerService class right now by running the code with unique code names for all components but this probably isn't optimal because it creates a new Code object every time instead of appending the new instance to the existing Code object.

**Comment by jungkumseok** [comment]
Fixed in 7f88ebd
